src/main/scala/com/scality/clueso/query/MetadataQueryExecutor.scala

```diff
@@ -0,0 +1,255 @@
+import org.apache.spark.SparkContext
+import org.apache.spark.clueso.metrics.SearchMetricsSource
+import org.apache.spark.sql.expressions.Window
+import org.apache.spark.sql.functions.{col, _}
+import org.apache.spark.sql.{DataFrame, Dataset, Row, SparkSession}
+
+import scala.concurrent.duration.Duration
+import scala.concurrent.{Await, ExecutionContext, Future, TimeoutException}
+
+class MetadataQueryExecutor(spark : SparkSession, config : CluesoConfig) extends LazyLogging {
+
+  import MetadataQueryExecutor._
+
+  implicit val _config = config
+  implicit val ec = ExecutionContext.fromExecutor(Executors.newFixedThreadPool(10))
+
+  var lockedBucketName : Option[String] = None
+
+  SearchMetricsSource(spark, config)
+
+  val metricsRegisterCancel = new AtomicBoolean(false)
+
+  val metricsRegisterThread = Future {
+    while (!metricsRegisterCancel.get()) {
+      // TODO change to config?
+      Thread.sleep(5000)
+
+//      logger.info("Registering new metrics")
+      SearchMetricsSource.registerRddMetrics(spark)
+//      logger.info("Done")
+    }
+  }
+
+  sys.addShutdownHook {
+    metricsRegisterCancel.set(true)
+  }
+
+
+  def getBucketDataframe(bucketName : String) : DataFrame = {
+    if (!config.cacheDataframes) {
+      logger.info("Cache is off.")
+      setupDf(spark, config, bucketName)
+    } else {
+      // delegate to cache manager
+      SessionCacheManager.getCachedBucketDataframe(spark, bucketName)
+    }
+  }
+
+  def printResults(f : Future[Array[String]], duration: Duration) = {
+    try {
+      val result = Await.result[Array[String]](f, duration)
+      println(s"RESULT:[${result.mkString(",")}]")
+    } catch {
+      case t:TimeoutException => {
+        println("")
+      }
+      case t:Throwable => {
+        logger.error("Error while printing results", t)
+        println(s"ERROR:${t.toString}")
+      }
+    }
+
+  }
+
+  def execAsync(query : MetadataQuery) = {
+    Future {
+      SparkUtils.getQueryResults(spark, this, query)
+    }
+  }
+
+  def executeAndPrint(query : MetadataQuery) = {
+    val resultArray = SparkUtils.getQueryResults(spark, this, query)
+    println("[" +  resultArray.mkString(",") + "]")
+  }
+
+  def execute(query : MetadataQuery) = {
+    val bucketName = query.bucketName
+
+    SearchMetricsSource.countSearches(1)
+
+    var resultDf = getBucketDataframe(bucketName)
+
+    val sqlWhereExpr = query.sqlWhereExpr
+    if (!sqlWhereExpr.isEmpty) {
+      resultDf = resultDf.where(sqlWhereExpr)
+    }
+
+    if (query.start_key.isDefined) {
+      resultDf = resultDf.where(col("key") > lit(query.start_key.get))
+    }
+
+    val currentWorkers = Math.max(currentActiveExecutors(spark.sparkContext).count(_ => true), 1)
+
+    val result = resultDf
+      .select(CluesoConstants.resultCols: _*)
+      .orderBy(col("key"))
+      .limit(query.limit)
+
+//    logger.info("Explain Query:")
+//    result.explain(true)
+//    logger.info("End Explain Query")
```

Comment:
The reason to have this commented is because the explain affects performance – explain is implemented in Spark as a ExplainCommand that gets executed across workers (think of it as a dry run). Because of this I've commented out but left in case it's needed for performance... Shall we wrap it in a config variable?  spark_sql_explains = true or false